src/NonceTracker.ts

```diff
@@ -0,0 +1,251 @@
+ *  @property opts.getPendingTransactions - A function that returns an array of txMeta
+ *  whose status is `submitted`
+ *  @property opts.getConfirmedTransactions - A function that returns an array of txMeta
+ *  whose status is `confirmed`
+ */
+export interface NonceTrackerOptions {
+  provider: Record<string, unknown>;
+  blockTracker: any;
+  getPendingTransactions: (address: string) => Transaction[];
+  getConfirmedTransactions: (address: string) => Transaction[];
+}
+
+/**
+ * @property highestLocallyConfirmed - A hex string of the highest nonce on a confirmed transaction.
+ * @property nextNetworkNonce - The next nonce suggested by the eth_getTransactionCount method.
+ * @property highestSuggested - The maximum between the other two, the number returned.
+ * @property local - Nonce details derived from pending transactions and highestSuggested
+ * @property network - Nonce details from the eth_getTransactionCount method
+ */
+export interface NonceDetails {
+  params: {
+    highestLocallyConfirmed: number;
+    nextNetworkNonce: number;
+    highestSuggested: number;
+  };
+  local: HighestContinuousFrom;
+  network: NetworkNextNonce;
+}
+
+/**
+ * @property nextNonce - The highest of the nonce values derived based on confirmed and pending transactions and eth_getTransactionCount method
+ * @property nonceDetails - details of nonce value derivation.
+ * @property releaseLock
+ */
+export interface NonceLock{
+  nextNonce: number;
+  nonceDetails: NonceDetails;
+  releaseLock: () => void;
+}
+
+/**
+ * @property name - The name for how the nonce was calculated based on the data used
+ * @property nonce - The next nonce value suggested by the eth_getTransactionCount method.
+ * @property blockNumber - The latest block from the network
+ * @property baseCount - Transaction count from the network suggested by eth_getTransactionCount method
+ */
+export interface NetworkNextNonce{
+  name: string;
+  nonce: number;
+  details: {
+    blockNumber: string;
+    baseCount: number;
+  };
+}
+
+/**
+ * @property name - The name for how the nonce was calculated based on the data used
+ * @property nonce - The next suggested nonce
+ * @property details{startPoint, highest} - the provided starting nonce that was used and highest derived from it (for debugging)
+ */
+export interface HighestContinuousFrom{
+  name: string;
+  nonce: number;
+  details: {
+    startPoint: number;
+    highest: number;
+  };
+}
+
+export interface Transaction {
+  status: string;
+  history: [Record<string, unknown>];
+  txParams: {
+    from: string;
+    gas: string;
+    value: string;
+    nonce: string;
+  };
+}
+
+export class NonceTracker {
+  private provider: Record<string, unknown>;
+
+  private blockTracker: any;
+
+  private ethQuery: any;
+
+  private getPendingTransactions: ((address: string) => Transaction[]);
+
+  private getConfirmedTransactions: ((address: string) => Transaction[]);
+
+  private lockMap: Record<string, unknown>;
```

Comment:
Does `await-semaphore` have types? If so, we should use `Mutex` or `typeof Mutex` instead of `unknown`.

If `await-semaphore` does not have types, we should define an interface for `Mutex` and use that:
```typescript
type ReleaseLock = () => void

interface AwaitSemaphoreMutex {
  acquire: () => ReleaseLock
}
```

If you define the `ReleaseLock` type, let's use that for all the `releaseLock` variables/properties/return values.

Reply:
Mutex was available as a type for the mutex. but the usage of interface AwaitSemaphoreMutex as mentioned was throwing error... So I proceeded with Mutex for returning mutex.